### DISCUSSION TITLE
PineTime Watch Support

### DISCUSSION THREAD

**Discussion by jones139** [OpenSeizureDetector/Android_Pebble_SD#70] [04_Devices]
Early in 2022 a user (@rennard) was working on making OSD run on the low cost [PineTime](https://pine64.com/product/pinetime-smartwatch-sealed/) watch using the MicroPython operating syststem.
This is really good because I want OpenSeizureDetector to work on very low cost hardware.   I have not managed to get it working on my PineTime yet though - need some more time to work out what is wrong!
His version is here if anyone wants to try it:  https://github.com/rennard/wasp-os-fork

**Comment by jones139** [comment]
I know someone else  (@mjones235) has also had a go at modifying the InfiniTime C++ operating system for PineTime to add OpenSeizureDetector support - again that will need testing to check the reliability.  
His version is here if anyone wants to try it:  https://github.com/MJones235/OSDInfiniTime

**Comment by jones139** [comment]
I have just ordered myself a couple more PineTime watches so I can get back to working on this, because I think it is important to help make OpenSeizureDetector accessible to people on low incomes, because the Garmin watches are expensive.

**Comment by jones139** [comment]
@44616E said:  "I like the look of a PineTime. There doesn't seem to be any dev kits available. I wouldn't mind have a crack at it. What path were you thinking of, did you plan to have an app to install on the watch or a dedicated firmware to make it a dedicated device. I would probablely be interested in the latter.I am concerned a little over long term hardware availability considering the delays in OS development."
You should not need a dev kit nowadays - I think the last time I used it I could flash it with different firmware without using the JTAG (or whatever it is) interface.    Just the risk of bricking it and not being able to recover if you mangle the bootloader I think.
When I set about a custom watch design I was intending to use a very simple freeRTOS based firmware for it - the approach of 'do one thing well' rather than trying to be a multi-purpose device.    I'd be quite happy to do that.   I was thinking of  piggy-backing on the InfiniTime firmware and modifying that - but it is much more complicated than we really need for OSD, so could just use their hardware driver libraries.   @mJones235 has modified InfiniTime to add OSD support here https://github.com/MJones235/OSDInfiniTime, but I think it needs updating for the latest InfiniTime version as it has changed a lot over the last year.    We probably need a bit of iteration between the Android APP BLE Data Source and the PineTime firmware to get it working (finalise the BLE characteristic IDs etc).

One advantage of doing our own simple firmware rather than linking too closely to infinitime would be that porting to different hardware would be easier maybe?

Graham

**Reply by 44616E** [reply]
We could also likely get better reliability and power consumption going low level. The problem is going to be with the HR sensor.

**Comment by jones139** [comment]
That is a good point - both the InfiniTime and WaspOS firmware have
implementations of a heart rate measurement algorithm, but I do not know
how good it is - the professional devices probably do some clever filtering
in software to get reliable readings - might need some more research to see
what is published on how best to do it.
I'd be quite interested to know how it is done, because I can tell if my
heart rhythm messes about when I am running, the indicated heart rate goes
up a lot - but I don't think that is real - more a measurement artefact.

On Thu, 27 Jul 2023 at 12:00, 44616E ***@***.***> wrote:

> We could also likely get better reliability and power consumption going
> low level. The problem is going to be with the HR sensor.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/OpenSeizureDetector/discussions/70#discussioncomment-6564051>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACLSYYVEGNZAC4WJDHMZKDXSJC5HANCNFSM6AAAAAATTELHO4>
> .
> You are receiving this because you authored the thread.Message ID:
> <OpenSeizureDetector/Android_Pebble_SD/repo-discussions/70/comments/6564051
> @github.com>
>


-- 
Graham Jones <http://google.com/+GrahamJones>
Hartlepool, UK.

**Reply by jones139** [reply]
I have started working on PineTime again so I can compare the heart rate monitor raw data to the values I am seeing on BangleJS (I am not convinced that the BangleJS HRM is working very well).
My initial intention was to try to get the micropython implementation by @rennard working - this is in the WaspOsSD repository now....but I am having a lot of trouble with memory errors from waspos, so I am wondering if trying to run python is just a bit too much of an ask for a small microcontroller.

In which case my option is to write it in C/C++ like Infinitime.

I shied away from the Infinitime approach last time I looked because it is written in 'modern' c++ and I was struggling to follow it (my comfort zone is old fashioned C).  Will have to try to educate myself I think....

**Comment by jones139** [comment]
I've just been reading through the Infinitime documentation about its [BLE implementation](https://github.com/InfiniTimeOrg/InfiniTime/blob/main/doc/ble.md).   It looks as though the latest versions of Infinitime provide BLE services/characteristics for heart rate, battery and xyz acceleration.   This should be enough to connect to OSD directly, if I add these services/characteristics into the Android App BLE data source.
I just have to re-flash ininitime onto my pinetime watch now without bricking it and having to prise the back off....

**Comment by jones139** [comment]
I was looking through the [Infinitime BLE Implementationdocumentation](https://github.com/InfiniTimeOrg/InfiniTime/blob/main/doc/ble.md) last night and noticed that the developers have already got it providing a heart rate and raw accelerometer data service.
This means that it should be pretty close to being usable with OpenSeizureDetector without major modifications, which is excellent news!

I'll update the V4.2.x branch of the main android app so it responds to the infinitime BLE services and see how it performs.

**Reply by jones139** [reply]
The Pinetime seems to be working pretty well now with the latest [V4.2.x version](https://github.com/OpenSeizureDetector/Android_Pebble_SD/tree/V4.2.x) of the OpenSeizureDetector android app, and the latest version of our custom PineTime Firmware, [PineTimeSD](https://github.com/OpenSeizureDetector/PineTimeSD).
There are a [couple of issues](https://github.com/OpenSeizureDetector/PineTimeSD/issues) with it, the most significant being to do with the rate we receive data, but I think I have a work around that I can easily implement in the Android App before release, even though I do not know what the real cause is.

**Reply by jones139** [reply]
I think that the major issue regarding receiving data too frequently has been resolved in V4.2.1l of the [OpenSeizureDetector Android App](https://github.com/OpenSeizureDetector/Android_Pebble_SD/tree/V4.2.x)

**Comment by jones139** [comment]
Comparison of Current (21 Jan 2024) version so of OpenSeizureDetector on PineTime and BangleJS, compared to Garmin devices.

| Requirement | PineTime | BangleJS2 | Garmin |
| ---       | ---              | ---               | ---           |
| Send Accelerometer Data to Phone at 25 Hz | Working (but may be slightly more than 25Hz) | Working | Working |
| Send Heart Rate Data to Phone at 0.2Hz | Working ok for normal movement - errors for large movements | Unreliable | Very Good |
| Report battery level to phone | Working OK | Working OK | Working OK |
| Mute Alarms from Watch | Not Yet Implemented | Not Yet Implemented |Working ok |
| Report Seizure Detector Status on Watch Screen | Not Yet Implemented | Not Yet Implemented |Working ok |
| Battery Life more than 12 hours | More than 24 hours | around 15 hours | Depends on watch - some are more than 24 hours, others are close to 12 hours |
| Ease of Installation | Tricky - install gadgetbridge, pair with watch, download firmware, disconnect gadgetbridge, pair with OSD | OK - update firmware from web app store | OK - copy firmware file onto watch via a computer |
| Cost (delivered to UK)        | ~ £35 (incl. delivery and taxes) | £76.80 | £179 (VenuSQ 2) |

Given the very promising results from PineTime (especially battery life and heart rate monitor performance) and the much lower cost than the other devices, I intend to concentrate on getting PineTime to being ready for release for general users.   The first priority will be to sort out some simple installation instructions, then improve the user interface.

**Reply by jones139** [reply]
I have been doing some tests with older android devices to see if we can get the PineTime + Android device combination down to less than £100.   
I have a Nokia 1 (which runs Android 10 Go Edition), and have used that to produce the step-by-step [user installation instructions](https://openseizuredetector.github.io/pages-user/pine-time-installation.html).
It seems to work ok, but I have noticed that the bluetooth connection reliability is not as good as I have seen on my usual development phone (a Samsung A20e running the full version of Android 11).

I wonder if this might be to do with the bluetooth version on the different devices:  [Pinetime](https://pine64.org/devices/pinetime/) supports Bluetooth V5, whereas the [Nokia 1](https://en.wikipedia.org/wiki/Nokia_1) is  Bluetooth V4.2, whereas the [Samsung A20e](https://www.gsmarena.com/samsung_galaxy_a20e-9661.php) is V5.

So we might have to recommend that the Android device supports Bluetooth V5 for best performance?  

It seems that the most relevant [difference between Bluetooth V4.2 and V5](https://www.symmetryelectronics.com/blog/bluetooth-5-versus-bluetooth-4-2-what-s-the-difference/) is much longer range.   Now the range of Bluetooth 4.2 should be fine for our application, but maybe it is just more susceptible to interference which is why my V5 device seems to be working better.    I'll do another test on my Samsung device, just in case I damaged the antenna when I disassembled and reassembled my test pinetime watch.

**Reply by jones139** [reply]
The PineTime seems to work very well with the Samsung A20e.   If you go out of bluetooth range and back it usually reconnects itself automatically.   If you are out of range for a long time (>1 hour, but not sure how long it has to be), it may fail to reconnect automatically but shutting down and re-starting the OSD phone app  gets it working ok.

If you would like to try the PineTime, please see the installation instructions here:  https://openseizuredetector.github.io/pages-user/pine-time-installation.html

(and feed back any experiences and suggest improvements to the installation process or instructions!)

**Comment by jeffrysleddens** [comment]
Hi! First up many thanks for all your hard work on this project! We have a son too with epilepsy and autism and it is so amazing to see a project like this. I'm hoping that I can at some point contribute too in some way.

I got the PineTime watch working with OpenSeizureDetector with version 4.2.1, but it seems that the watch goes into a sleepmode or something and then stops communicating with the phone and then the app goes in to FAULT state. I'm wondering if there is some settings that I missed. The PineTime watch is running InifiTime version 1.13.0 and I did not change any settings on the watch itself.

**Comment by jones139** [comment]
Hi,
You need to run a custom version of infinitime that I have created to
address that issue and also make it sample the accelerometer readings fast
enough for the seizure detection algorithm.

Please see the installation instructions here:
https://openseizuredetector.github.io/pages-user/pine-time-installation.html

Let me know how you get on!

Thanks

Graham ***@***.***)

**Reply by jeffrysleddens** [reply]
Ah, I can't believe I missed that!  The instructions on that page were super clear and I was able to install the OSD firmware onto the watch without any issues. It looks like it is all working perfectly now! Again, thank you very much!

**Reply by jones139** [reply]
Thanks, I am glad the instructions worked - I need to decide whether to put the effort into making the firmware update part of OSD rather than relying on Gadgetbridge.   A few users have managed to do it ok using the GadgetBridge instructions, but I suspect they may be the sort of users who are happy to fiddle to get it to work, and others may not!

**Reply by lundstrj** [reply]
What a wonderfully well made guide for the PineTime!

**Comment by lundstrj** [comment]
About the PineTime.
Yesterday I installed PineTimeSD 0.6 and ran that with OSD Android 4.2.3 (release apk downloaded from github) and unfortunately it was a bit of a rocky night.

I got multiple faults where OSD Android reported the watch app as NOT running. At the same time it was still updating accelerometer data and heart rate (I observed it for a good solid 20-30 seconds) so PineTimeSD was clearly was running and sending data.

I think I'll try 4.2.2, or perhaps build the latest 4.2.x branch to see if I get better results.

(sorry about the giant screenshots)

![Screenshot_20240325_022920_OpenSeizureDetector](https://github.com/OpenSeizureDetector/Android_Pebble_SD/assets/1045735/cc4ca22f-3df9-4327-9536-da7d0f58cae5)
![Screenshot_20240325_022922_OpenSeizureDetector](https://github.com/OpenSeizureDetector/Android_Pebble_SD/assets/1045735/fa5f156a-b42d-45ba-8ab7-1ad19146bccb)
![Screenshot_20240325_022926_OpenSeizureDetector](https://github.com/OpenSeizureDetector/Android_Pebble_SD/assets/1045735/3edd3b58-7c27-445f-8cb7-251102df9816)

**Reply by jones139** [reply]
Thanks for reporting this @lundstrj.   The time in the OSD app is 02:27:03, but the phone time is 02:29.  The time in the app is the last time it received a 5 second data sample from the watch, so it looks like the communications did stall, so the FAULT condition is correct.

I did wonder if you have got the system into the state where there are two versions of the OSD server running at once (when that is the case it alternates between OK and FAULT) - a force-stop of OSD or re-boot of the phone usually fixes that.  The issue is related to stopping and starting the system - sometimes it does not shutdown properly.  Do you think this is a possibility for your issue?

**Reply by lundstrj** [reply]
I mean, it is not impossible at all that it is how you suggest. I did find it odd that the displayed heart rate was updating and when I was gently shaking the watch the accelerometer graph shot up. It was like the system was in a weird state of limbo where data was flowing but OSD still warned me that the connection was broken. I haven't looked into what determines if the watch is connected or not. Is it at all possible for OSD to detect the watch as disconnected even though data is being received?

**Reply by jones139** [reply]
It does not really do watch connection checks - it is really a 'data received on time' check.  I'll wear the watch tonight and watch more carefully what happens when I brush my teeth, which always sets it off.
